File: src/core/engine/machine.ts
/**
 * The interpreter: step 6 of the bar cycle, and the whole architecture of the
 * language in one loop.
 *
 * **It walks a list. There is no `eval`, no generated function and no code
 * built from text anywhere in it**, which is what lets the language run inside
 * an application under a content security policy that forbids turning text into
 * code, and what lets a platform run many customers' scripts in one process.
 * The compiler emits data and this walks it; that is the foundation of the
 * whole design rather than a preference.
 *
 * **What verification already proved, this does not re-check.** A verified
 * program cannot underflow the stack, cannot jump out of bounds, cannot address
 * a slot that does not exist and cannot loop without charging the budget (3.5).
 * So the loop below pops without testing for an empty stack and indexes without
 * testing a bound, and every remaining failure is a script error with a source
 * position, which is the only kind of failure a user should ever see.
 *
 * **The budget is checked on the path every instruction takes.** One increment
 * and one comparison, because the alternative is a loop an engine hopes will
 * end. The loop budget itself is charged by `TICK` and by nothing else, so two
 * engines run out at the same iteration of the same loop on the same bar (8.5).
 */
import type { Span } from '../span/index.js';
import {
  add,
  and,
  compare,
  divide,
  equals,
  isFalsey,
  multiply,
  negate,
  not,
  or,
  remainder,
  subtract,
} from './arithmetic.js';
import type { Budget } from './budget.js';
import type { Channels } from './channels.js';
import type { Frame } from './frames.js';
import { makeFrame, positionAt } from './frames.js';
import type {
  BarView,
  CallContext,
  Guard,
  HostFacts,
  ManifestEntry,
  PositionFacts,
  StateRecord,
} from './library/index.js';
import type { Memory } from './memory.js';
import {
  constantValueOf,
  elementRead,
  forInit,
  forNext,
  historyRead,
  nameOf,
} from './operations.js';
import type { Ops } from './operations.js';
import type { Registers } from './registers.js';
import type { CompiledProgram, Instruction, Position } from './types.js';
import type { Heap, Value } from './values/index.js';
import { ABSENT, reference } from './values/index.js';

/** Everything the interpreter is handed once, at load. */
export interface MachineParts {
  readonly program: CompiledProgram;
  readonly registers: Registers;
  readonly memory: Memory;
  readonly channels: Channels;
  readonly heap: Heap;
  readonly budget: Budget;
  readonly guard: Guard;
  readonly host: HostFacts;
  /** The strategy's own position, read from the run's ledger. */
  readonly position: PositionFacts;
  /** The library entry behind each `lib.functions` index, resolved at load. */
  readonly library: readonly ManifestEntry[];
  /** The position table of each function body, by function index. */
  readonly fnPositions: readonly (readonly Position[])[];
  spanAt(line: number, column: number): Span;
}

/**
 * The region a stateless call is handed.
 *
 * Shared rather than allocated per call, because an entry declared without
 * state never touches it and an allocation on the path of every `max` and every
 * colour would be the engine's largest per bar cost for nothing.
 */
const NO_STATE: StateRecord = {};

/**
 * The reused call context.
 *
 * `CallContext` declares `state` read-only, because a library call is handed a
 * region rather than choosing one. The machine is the one place that sets it,
 * so it holds the same object through a type that can.
 */
interface MutableContext extends Omit<CallContext, 'state'> {
  state: StateRecord;
}

export class Machine {
  private readonly parts: MachineParts;
  private readonly top: Frame;
  private readonly callers: Frame[] = [];
  private frame: Frame;
  private view: BarView;
  private barIndex = 0;
  /** The loop whose `TICK` executed most recently, for OS5001's line. */
  private lastLoop = -1;
  /**
   * The value a top level `RET` handed back, `compiled-program.md` 2.16.1.
   *
   * A program's own `code` ends in `HALT` and produces nothing, so this stays
   * absent for it. A request body ends in `RET` and its value is the read's
   * value for that requested bar, which is the one thing the interpreter has to
   * hand out rather than write into a register or a channel.
   */
  private returned: Value = ABSENT;
  private readonly ctx: MutableContext;
  private readonly ops: Ops;

  constructor(parts: MachineParts, view: BarView) {
    this.parts = parts;
    this.view = view;
    const { program } = parts;
    this.top = makeFrame(program.code, program.frame.slots, 0, 0, [], program.debug.pos);
    this.frame = this.top;
    this.ops = this.operations();

    // One context, reused by every call, with the two fields that change per
    // call read through getters. A fresh object per `CALL_LIB` would be an
    // allocation on the hottest path in the engine, and the span is a binary
    // search that most calls never ask for.
    const machine = this;
    this.ctx = {
      heap: parts.heap,
      host: parts.host,
      position: parts.position,
      guard: parts.guard,
      state: NO_STATE,
      get span(): Span {
        return machine.here();
      },
      get bar(): BarView {
        return machine.view;
      },
      nameOf: (value: Value): string => machine.nameOf(value),
    };
  }

  /** Step 3: the stack, the slots and the frame stack all start empty. */
  begin(view: BarView, index: number): void {
    this.view = view;
    this.barIndex = index;
    this.lastLoop = -1;
    this.returned = ABSENT;
    this.callers.length = 0;
    this.frame = this.top;
    this.top.pc = 0;
    this.top.stack.length = 0;
    this.top.slots.fill(ABSENT);
  }

  /** What the last `run` handed back, for an instruction list that ends in `RET`. */
  result(): Value {
    return this.returned;
  }

  /** The slot array of frame 0, which inputs and grids are written into. */
  slots(): Value[] {
    return this.top.slots;
  }

  /** Where the instruction that just executed came from. */
  here(): Span {
    const at = positionAt(this.frame.positions, Math.max(this.frame.pc - 1, 0));
    return this.parts.spanAt(at.line, at.column);
  }

  /**
   * Runs the bar.
   *
   * The dispatch is a switch over the opcode name. A name rather than a number
   * is what the format carries (2.14) so that a program is readable and
   * diffable by hand; an engine may map names to its own integers at load and
   * must not depend on any numbering, because none is defined.
   */
  run(): void {
    const { budget, channels, memory, registers, heap } = this.parts;
    for (;;) {
      if (budget.step()) budget.overrun(this.here(), this.loopLine());
      const frame = this.frame;
      const instruction = frame.code[frame.pc] as Instruction;
      frame.pc += 1;
      const stack = frame.stack;

      switch (instruction[0]) {
        case 'CONST':
          stack.push(constantValueOf(this.parts.program.consts[instruction[1] as number]));
          break;
        case 'DUP':
          stack.push(stack[stack.length - 1] ?? ABSENT);
          break;
        case 'POP':
          stack.pop();
          break;
        case 'LOAD':
          stack.push(frame.slots[instruction[1] as number] ?? ABSENT);
          break;
        case 'STORE':
          frame.slots[instruction[1] as number] = pop(stack);
          break;
        case 'CELL_INIT':
          if (memory.initialise(frame.cellBase + (instruction[1] as number))) {
            frame.pc = instruction[2] as number;
          }
          break;
        case 'LOADC':
          stack.push(memory.load(frame.cellBase + (instruction[1] as number)));
          break;
        case 'STOREC':
          memory.store(frame.cellBase + (instruction[1] as number), pop(stack));
          break;
        case 'SLOAD':
          stack.push(registers.get(instruction[1] as number));
          break;
        case 'SSTORE':
          registers.set(instruction[1] as number, pop(stack));
          break;
        case 'HIST':
          stack.push(historyRead(this.ops, instruction[1] as number, pop(stack)));
          break;
        case 'HISTP':
          stack.push(historyRead(this.ops, frame.series[instruction[1] as number] ?? -1, pop(stack)));
          break;
        case 'ADD':
          stack.push(this.joined(stack));
          break;
        case 'SUB':
          stack.push(binary(stack, subtract));
          break;
        case 'MUL':
          stack.push(binary(stack, multiply));
          break;
        case 'DIV':
          stack.push(binary(stack, divide));
          break;
        case 'MOD':
          stack.push(binary(stack, remainder));
          break;
        case 'NEG':
          stack.push(negate(pop(stack)));
          break;
        case 'LT':
        case 'LE':
        case 'GT':
        case 'GE': {
          const b = pop(stack);
          stack.push(compare(instruction[0], pop(stack), b));
          break;
        }
        case 'EQ': {
          const b = pop(stack);
          stack.push(equals(pop(stack), b));
          break;
        }
        case 'NE': {
          const b = pop(stack);
          stack.push(!equals(pop(stack), b));
          break;
        }
        case 'NOT':
          stack.push(not(pop(stack)));
          break;
        case 'AND':
          stack.push(binary(stack, and));
          break;
        case 'OR':
          stack.push(binary(stack, or));
          break;
        case 'AND_SHORT':
          if (stack[stack.length - 1] === false) frame.pc = instruction[1] as number;
          break;
        case 'OR_SHORT':
          if (stack[stack.length - 1] === true) frame.pc = instruction[1] as number;
          break;
        case 'JUMP':
          frame.pc = instruction[1] as number;
          break;
        case 'JUMP_FALSE':
          if (isFalsey(pop(stack))) frame.pc = instruction[1] as number;
          break;
        case 'TICK': {
          const loop = instruction[1] as number;
          this.lastLoop = loop;
          budget.tick(this.here(), this.parts.program.loops[loop]?.line ?? 0);
          break;
        }
        case 'FOR_INIT':
          forInit(this.ops, instruction, frame);
          break;
        case 'FOR_NEXT':
          forNext(instruction, frame);
          break;
        case 'ARRAY': {
          const count = instruction[1] as number;
          const items = stack.splice(stack.length - count, count);
          this.parts.guard.array(this.here(), 'the array', items.length);
          stack.push(reference(heap.allocate({ kind: 'array', items })));
          break;
        }
        case 'ELEM': {
          const index = pop(stack);
          stack.push(elementRead(this.ops, pop(stack), index));
          break;
        }
        case 'CALL_LIB':
          this.callLibrary(instruction, stack);
          break;
        case 'CALL_FN':
          this.callFunction(instruction[1] as number, stack);
          break;
        case 'RET': {
          const value = pop(stack);
          const caller = this.callers.pop();
          if (caller === undefined) {
            this.returned = value;
            return;
          }
          this.frame = caller;
          caller.stack.push(value);
          break;
        }
        case 'EMIT':
          channels.write(instruction[1] as number, pop(stack));
          break;
        case 'HALT':
          return;
        default:
          // Unreachable in a verified program: check 2 refused every opcode
          // this switch does not have an arm for.
          throw new Error(`${String(instruction[0])} is not an instruction`);
      }
    }
  }

  private loopLine(): number | undefined {
    if (this.lastLoop < 0) return undefined;
    return this.parts.program.loops[this.lastLoop]?.line;
  }

  /** What the operations in `operations.ts` need, built once. */
  private operations(): Ops {
    const machine = this;
    return {
      program: this.parts.program,
      registers: this.parts.registers,
      memory: this.parts.memory,
      heap: this.parts.heap,
      slots: this.top.slots,
      here: () => machine.here(),
      barIndex: () => machine.barIndex,
    };
  }

  /** `ADD` with the string ceiling applied, since it is the one that grows one. */
  private joined(stack: Value[]): Value {
    const b = pop(stack);
    const value = add(pop(stack), b);
    return typeof value === 'string' ? this.parts.guard.string(this.here(), value) : value;
  }

  /**
   * `CALL_LIB`.
   *
   * A call whose manifest entry carries an effect does not perform it: it
   * appends a record holding the function index and the argument values as they
   * stood, and pushes absent. The machine does that here rather than in the
   * entry, so an effect cannot be forgotten by whoever writes the next one.
   */
  private callLibrary(instruction: Instruction, stack: Value[]): void {
    const index = instruction[1] as number;
    const count = instruction[2] as number;
    const state = instruction[3] as number;
    const args = stack.splice(stack.length - count, count);
    const of = this.parts.library[index] as ManifestEntry;

    if (of.effect !== 'none') {
      this.parts.channels.defer({ fn: index, name: of.name, effect: of.effect, args });
      stack.push(ABSENT);
      return;
    }

    this.ctx.state =
      state < 0 ? NO_STATE : this.parts.memory.region(this.frame.stateBase + state);
    stack.push(of.call(this.ctx, args));
  }

  /** `CALL_FN`: everything it needs is in the call site, 4.10. */
  private callFunction(site: number, stack: Value[]): void {
    const { program, fnPositions } = this.parts;
    const call = program.callSites[site];
    if (call === undefined) return;
    const target = program.functions[call.fn];
    if (target === undefined) return;
    const args = stack.splice(stack.length - call.argc, call.argc);
    const frame = makeFrame(
      target.code,
      target.slots,
      call.cellBase,
      call.stateBase,
      call.series,
      fnPositions[call.fn] ?? [],
    );
    for (let i = 0; i < args.length; i += 1) frame.slots[i] = args[i] ?? ABSENT;
    this.callers.push(this.frame);
    this.frame = frame;
  }

  /** The name a value goes by, which a library call asks for by the same route. */
  nameOf(value: Value): string {
    return nameOf(this.ops, value);
  }
}

/** The stack cannot be empty here: the verifier's depth walk proved it. */
function pop(stack: Value[]): Value {
  return stack.pop() ?? ABSENT;
}

function binary(stack: Value[], of: (a: Value, b: Value) => Value): Value {
  const b = pop(stack);
  return of(pop(stack), b);
}

File: engine/openscript/machine.py
"""The instruction loop: forty-one arms and nothing clever in any of them.

Section 4 is deliberately small and deliberately dull. There is one way to do
each thing, no instruction is a shorthand for two others, and a new one cannot be
added without a major format bump. So this is a loop over a list, with a table of
handlers, and the table is data: a name in a compiled program never becomes a
function here, and nothing on this path builds anything out of text.

**Verification is what lets this be plain.** A verified program cannot underflow
the stack, cannot jump out of bounds, cannot address a slot that does not exist
and cannot loop without charging the budget (section 3.5), so none of that is
tested again per instruction. What is tested per instruction is the one thing
verification cannot decide: the tags of the values, which is where absence lives.

**Absence has a rule in every arm and the rule is the page's.** It propagates
through arithmetic and ordering comparison, is total under equality, is three
valued under the logical operators, and is false at a branch. Those five rules
are ``values.py``'s, and an engine that special cased absence anywhere else would
have a bug rather than an optimisation.
"""

import math
from typing import Any, Callable, Dict, List, Optional, Sequence

from .budget import Budget
from .contracts import CallContext, Library
from .diagnostics import NO_POSITION, Position, raise_at
from .memory import Cells, Channels, Frame, Register, States, read_history
from .program import LoadedProgram
from .values import (
    ABSENT,
    ArrayValue,
    conjunction,
    disjunction,
    equal,
    finite,
    is_number,
    is_whole,
    negation,
    ordered,
    stored,
    truthy,
)


class PendingEffect:
    """One library call that reaches the outside world, held until step 9.

    A function with an effect does not perform it when it executes: it leaves a
    record here and pushes absent. The list is discarded at step 3 of every
    execution and applied at step 9 only when the bar is decided, which is what
    makes a condition that was true halfway through a bar and false when it
    closed place no order at all.
    """

    __slots__ = ("fn", "name", "effect", "arguments", "position")

    def __init__(
        self, fn: int, name: str, effect: str, arguments: List[Any], position: Position
    ) -> None:
        self.fn = fn
        self.name = name
        self.effect = effect
        self.arguments = arguments
        self.position = position


class Machine:
    """One program, its memory, and the loop that walks its instructions."""

    def __init__(
        self,
        program: LoadedProgram,
        cells: Cells,
        states: States,
        registers: List[Register],
        channels: Channels,
        budget: Budget,
        library: Library,
    ) -> None:
        self.program = program
        self.cells = cells
        self.states = states
        self.registers = registers
        self.channels = channels
        self.budget = budget
        self.library = library
        self.pending: List[PendingEffect] = []
        self.context = CallContext()
        self.bar_index = 0
        self._frames: List[Frame] = []
        self._loop_line: Optional[int] = None
        self._answer: Any = ABSENT
        self._running = False
        self._arms: Dict[str, Callable[[Frame, Sequence[Any]], None]] = {
            "CONST": self._const,
            "DUP": self._dup,
            "POP": self._pop_top,
            "LOAD": self._load,
            "STORE": self._store,
            "CELL_INIT": self._cell_init,
            "LOADC": self._load_cell,
            "STOREC": self._store_cell,
            "SLOAD": self._series_load,
            "SSTORE": self._series_store,
            "HIST": self._history,
            "HISTP": self._history_parameter,
            "ADD": self._add,
            "SUB": self._subtract,
            "MUL": self._multiply,
            "DIV": self._divide,
            "MOD": self._modulo,
            "NEG": self._negate,
            "LT": self._less,
            "LE": self._less_equal,
            "GT": self._greater,
            "GE": self._greater_equal,
            "EQ": self._equals,
            "NE": self._differs,
            "NOT": self._not,
            "AND": self._and,
            "OR": self._or,
            "AND_SHORT": self._and_short,
            "OR_SHORT": self._or_short,
            "JUMP": self._jump,
            "JUMP_FALSE": self._jump_false,
            "TICK": self._tick,
            "FOR_INIT": self._for_init,
            "FOR_NEXT": self._for_next,
            "ARRAY": self._array,
            "ELEM": self._element,
            "CALL_LIB": self._call_library,
            "CALL_FN": self._call_function,
            "RET": self._return,
            "EMIT": self._emit,
            "HALT": self._halt,
        }

    # -- the loop ---------------------------------------------------------

    def begin(self, bar_index: int, context: CallContext) -> None:
        """Step 3's share: a fresh frame 0 with every slot absent, and the counters.

        The frame exists before step 6 rather than at it, because step 5 writes
        each input's effective value into a slot of this frame and the two steps
        are in that order.
        """
        self.bar_index = bar_index
        self.context = context
        self.pending = []
        self._loop_line = None
        self.budget.begin()
        top = self.program.top
        self._frames = [Frame(top.code, top.slots, positions=top.positions)]

    def write_slot(self, slot: int, value: Any) -> None:
        """Step 5: one input's effective value, into frame 0."""
        self._frames[0].slots[slot] = stored(value)

    def execute(self) -> Any:
        """Step 6: run the top-level list from instruction 0 until ``HALT``."""
        return self._walk()

    def _walk(self) -> Any:
        self._answer = ABSENT
        self._running = True
        while self._running:
            current = self._frames[-1]
            if self.budget.step():
                self.budget.spent(self.here(), self._loop_line)
            instruction = current.code[current.pc]
            current.pc += 1
            self._arms[instruction[0]](current, instruction)
        return self._answer

    def here(self) -> Position:
        """Where the instruction that is executing came from in the source."""
        if not self._frames:
            return NO_POSITION
        frame = self._frames[-1]
        at = frame.pc - 1
        positions = frame.positions
        if positions is None or at < 0 or at >= len(positions):
            return NO_POSITION
        return positions[at]

    # -- constants and stack ----------------------------------------------

    def _const(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.stack.append(self.program.consts[instruction[1]])

    def _dup(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.stack.append(frame.stack[-1])

    def _pop_top(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.stack.pop()

    # -- slots -------------------------------------------------------------

    def _load(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.stack.append(frame.slots[instruction[1]])

    def _store(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.slots[instruction[1]] = stored(frame.stack.pop())

    # -- cells --------------------------------------------------------------

    def _cell_init(self, frame: Frame, instruction: Sequence[Any]) -> None:
        cell = frame.cell_base + instruction[1]
        if self.cells.initialised(cell):
            frame.pc = instruction[2]
            return
        # Marked before the initialiser runs, not after: the two differ only if
        # the initialiser could reach this instruction again, which needs
        # recursion, which is an error.
        self.cells.mark(cell)

    def _load_cell(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.stack.append(self.cells.read(frame.cell_base + instruction[1]))

    def _store_cell(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self.cells.write(frame.cell_base + instruction[1], frame.stack.pop())

    # -- series registers ---------------------------------------------------

    def _series_load(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.stack.append(self.registers[instruction[1]].current)

    def _series_store(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self.registers[instruction[1]].current = stored(frame.stack.pop())

    def _read_back(self, frame: Frame, register: int) -> None:
        back = frame.stack.pop()
        frame.stack.append(
            read_history(
                self.registers[register],
                back,
                self.bar_index,
                self.program.retained,
                self.here(),
            )
        )

    def _history(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._read_back(frame, instruction[1])

    def _history_parameter(self, frame: Frame, instruction: Sequence[Any]) -> None:
        # A parameter whose call site bound -1 cannot be reached by a HISTP: the
        # compiler binds a register only for a parameter whose history the body
        # reads, and a HISTP against -1 is a corrupt program the verifier caught.
        self._read_back(frame, frame.bindings[instruction[1]])

    # -- arithmetic ---------------------------------------------------------

    def _two(self, frame: Frame):
        right = frame.stack.pop()
        return frame.stack.pop(), right

    def _add(self, frame: Frame, instruction: Sequence[Any]) -> None:
        left, right = self._two(frame)
        if left is ABSENT or right is ABSENT:
            frame.stack.append(ABSENT)
            return
        if is_number(left) and is_number(right):
            frame.stack.append(finite(left + right))
            return
        if isinstance(left, str) and isinstance(right, str):
            # The one operator that grows a string, and so the one that meets
            # the ceiling: a log appended to on every bar is what OS5008 is for.
            frame.stack.append(self.budget.text(self.here(), left + right))
            return
        # Any other combination is a program the checker should have rejected,
        # and an engine that invented a conversion here would become the only
        # engine that runs that script.
        frame.stack.append(ABSENT)

    def _numeric(self, frame: Frame, of: Callable[[float, float], Any]) -> None:
        left, right = self._two(frame)
        if not is_number(left) or not is_number(right):
            frame.stack.append(ABSENT)
            return
        frame.stack.append(of(float(left), float(right)))

    def _subtract(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._numeric(frame, lambda a, b: finite(a - b))

    def _multiply(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._numeric(frame, lambda a, b: finite(a * b))

    def _divide(self, frame: Frame, instruction: Sequence[Any]) -> None:
        # Absent when the divisor is zero, zero divided by zero included: the
        # language has no infinity and no not-a-number to hand back.
        self._numeric(frame, lambda a, b: ABSENT if b == 0 else finite(a / b))

    def _modulo(self, frame: Frame, instruction: Sequence[Any]) -> None:
        # The remainder of truncated division, so the sign follows the left
        # operand rather than the divisor.
        self._numeric(frame, lambda a, b: ABSENT if b == 0 else finite(math.fmod(a, b)))

    def _negate(self, frame: Frame, instruction: Sequence[Any]) -> None:
        value = frame.stack.pop()
        frame.stack.append(finite(-float(value)) if is_number(value) else ABSENT)

    # -- comparison ---------------------------------------------------------

    def _order(self, frame: Frame, wanted: Callable[[float], bool]) -> None:
        left, right = self._two(frame)
        found = ordered(left, right)
        frame.stack.append(ABSENT if found is ABSENT else wanted(found))

    def _less(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._order(frame, lambda found: found < 0)

    def _less_equal(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._order(frame, lambda found: found <= 0)

    def _greater(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._order(frame, lambda found: found > 0)

    def _greater_equal(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._order(frame, lambda found: found >= 0)

    def _equals(self, frame: Frame, instruction: Sequence[Any]) -> None:
        left, right = self._two(frame)
        frame.stack.append(equal(left, right))

    def _differs(self, frame: Frame, instruction: Sequence[Any]) -> None:
        left, right = self._two(frame)
        frame.stack.append(not equal(left, right))

    # -- logic ---------------------------------------------------------------

    def _not(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.stack.append(negation(frame.stack.pop()))

    def _and(self, frame: Frame, instruction: Sequence[Any]) -> None:
        left, right = self._two(frame)
        frame.stack.append(conjunction(left, right))

    def _or(self, frame: Frame, instruction: Sequence[Any]) -> None:
        left, right = self._two(frame)
        frame.stack.append(disjunction(left, right))

    def _and_short(self, frame: Frame, instruction: Sequence[Any]) -> None:
        # The value that decides the answer by itself stays on the stack as the
        # answer. Absence decides neither operator, so it does not short circuit.
        if frame.stack[-1] is False:
            frame.pc = instruction[1]

    def _or_short(self, frame: Frame, instruction: Sequence[Any]) -> None:
        if frame.stack[-1] is True:
            frame.pc = instruction[1]

    # -- branching and loops --------------------------------------------------

    def _jump(self, frame: Frame, instruction: Sequence[Any]) -> None:
        frame.pc = instruction[1]

    def _jump_false(self, frame: Frame, instruction: Sequence[Any]) -> None:
        if not truthy(frame.stack.pop()):
            frame.pc = instruction[1]

    def _tick(self, frame: Frame, instruction: Sequence[Any]) -> None:
        line = self.program.loop_lines[instruction[1]]
        self._loop_line = line
        self.budget.tick(self.here(), line)

    def _for_init(self, frame: Frame, instruction: Sequence[Any]) -> None:
        step = frame.stack.pop()
        limit = frame.stack.pop()
        start = frame.stack.pop()
        bounds = ((start, "start"), (limit, "limit"), (step, "step"))
        for value, bound in bounds:
            if value is ABSENT:
                # An absent bound means the number of iterations is unknown, and
                # running zero times would hide that.
                raise_at("OS4013", self.here(), bound=bound)
        for value, _bound in bounds:
            if not is_number(value):
                raise_at("OS4001", self.here(), index=str(value))
        if step == 0:
            raise_at(
                "OS3004",
                self.here(),
                name="this loop",
                argument="step",
                range="anything but zero",
                found=0,
            )
        frame.slots[instruction[2]] = start
        frame.slots[instruction[3]] = limit
        frame.slots[instruction[4]] = step
        # A descending range with a positive step runs zero times and is never
        # silently reversed.
        if (start > limit) if step > 0 else (start < limit):
            frame.pc = instruction[5]

    def _for_next(self, frame: Frame, instruction: Sequence[Any]) -> None:
        limit = frame.slots[instruction[3]]
        step = frame.slots[instruction[4]]
        moved = finite(float(frame.slots[instruction[2]]) + float(step))
        frame.slots[instruction[2]] = moved
        if moved is ABSENT:
            return
        if (moved <= limit) if step > 0 else (moved >= limit):
            frame.pc = instruction[5]

    # -- arrays ----------------------------------------------------------------

    def _array(self, frame: Frame, instruction: Sequence[Any]) -> None:
        count = instruction[1]
        held = frame.stack[len(frame.stack) - count :] if count else []
        del frame.stack[len(frame.stack) - count :]
        # A new array every time it executes, which is why an array literal is
        # not a constant pool entry.
        frame.stack.append(ArrayValue(list(held)))

    def _element(self, frame: Frame, instruction: Sequence[Any]) -> None:
        index = frame.stack.pop()
        array = frame.stack.pop()
        size = len(array.elements) if isinstance(array, ArrayValue) else 0
        usable = (
            isinstance(array, ArrayValue)
            and is_whole(index)
            and 0 <= index < size
        )
        if not usable:
            # The opposite of a history read past the start of the dataset, and
            # the difference is the point: an array has an extent the script
            # chose, so an index outside it is a mistake.
            raise_at(
                "OS4004",
                self.here(),
                index="none" if index is ABSENT else str(index),
                name="the array",
                size=size,
            )
        frame.stack.append(array.elements[int(index)])

    # -- calls -------------------------------------------------------------------

    def _call_library(self, frame: Frame, instruction: Sequence[Any]) -> None:
        which, count, state = instruction[1], instruction[2], instruction[3]
        arguments = frame.stack[len(frame.stack) - count :] if count else []
        del frame.stack[len(frame.stack) - count :]
        entry = self.program.entries[which]
        if entry.effect != "none":
            self.pending.append(
                PendingEffect(which, entry.name, entry.effect, list(arguments), self.here())
            )
            frame.stack.append(ABSENT)
            return
        region = None if state < 0 else self.states.region(frame.state_base + state)
        # The string ceiling, both halves: the length a call can be asked for
        # before a character of it exists, and the string it did build. A call of
        # no arguments answers a fact the host stated rather than a string this
        # bar grew, and the ceiling is on what a script grows.
        self.budget.measured(self.here(), self.library.length_of(entry.name, arguments))
        answer = self.library.call(entry.name, list(arguments), region, self.context)
        if arguments and isinstance(answer, str):
            self.budget.text(self.here(), answer)
        frame.stack.append(stored(answer))

    def _call_function(self, frame: Frame, instruction: Sequence[Any]) -> None:
        site = self.program.call_sites[instruction[1]]
        body = self.program.functions[site["fn"]]
        count = site["argc"]
        arguments = frame.stack[len(frame.stack) - count :] if count else []
        del frame.stack[len(frame.stack) - count :]
        called = Frame(
            body.code,
            body.slots,
            cell_base=site["cellBase"],
            state_base=site["stateBase"],
            bindings=site["series"],
            positions=body.positions,
        )
        for at, value in enumerate(arguments):
            called.slots[at] = value
        self._frames.append(called)

    def _return(self, frame: Frame, instruction: Sequence[Any]) -> None:
        value = frame.stack.pop()
        self._frames.pop()
        if not self._frames:
            self._answer = value
            self._running = False
            return
        self._frames[-1].stack.append(value)

    # -- output and termination -----------------------------------------------------

    def _emit(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self.channels.write(instruction[1], frame.stack.pop())

    def _halt(self, frame: Frame, instruction: Sequence[Any]) -> None:
        self._answer = ABSENT
        self._running = False
